Fundamentals of regions are in.  Also, a lot of new JSON files.

jmgoldb opened the pull request:
Making a pull request early so we can get these JSON files.  Next I am going to set up the distinction of low, high, and g-rank region views, by subclassing the view controller I made.  I considered instead just having one view controller able to handle everything, but that would lead to needless complexity.

jmgoldb committed:
Made some progress into region navigation

jmgoldb committed:
Finished rudimentary low-rank region handling.  Still need to tweak the UI, and add in functionality for low-high-g-rank picker

jmgoldb committed:
Merge commit 'd13bcf7fbad9d3a54b9f498c09e64678c4a7aaaa'

jmgoldb committed:
Finalized Data for AreaDrops/regions

jmgoldb committed:
Addressed some comments in the 1st pull request

vokal-isaac commented on MH4U Dex/AreaDropTableViewCell.h:
```diff
@@ -0,0 +1,17 @@
+
+#import <UIKit/UIKit.h>
+
+#import "AreaDrop.h"
```
It's usually preferable to forward-declare the class in the header file (`@class AreaDrop;`), rather than importing that class's header, unless there's something more than the class symbol that's necessary for this header.

designatednerd replied:
(you do still have to `#import` in the `.m` file though)

jmgoldb replied:
So would that look something like:

#import "AreaDrop.m"

@class AreaDrop;

vokal-isaac replied:
Depending on how it's used in the implementation file, it may or may not be necessary to add `#import "AreaDrop.h"` in `AreaDropTableViewCell.m`—Xcode generally gives an error talking about forward-declarations if you try to do something with the class that requires the import.

vokal-isaac commented on MH4U Dex/CoreDataController.h:
```diff
@@ -18,6 +18,8 @@
 - (void)loadRegionData;
 - (BOOL)attemptSaveContext;
 
+- (NSManagedObject *)getUniqueEntityWithEntityName:(NSString *)entityName withPredicate:(NSPredicate *)predicate;
```
This (and, actually, most if not all methods exposed in headers) should have a documenting comment.

vokal-isaac commented on MH4U Dex/CoreDataController.m:
```diff
@@ -130,7 +130,7 @@ - (void)loadMonsterDropData
         // Only add the drop if it is not already there
         NSNumber *monsterDropID = drop[MonsterDropIDKey];
-        NSPredicate *monsterDropPredicate = [NSPredicate predicateWithFormat:@"%K == %d", MonsterDropIDKey, monsterDropID.intValue];
+        NSPredicate *monsterDropPredicate = [NSPredicate predicateWithFormat:@"%K == %d", MonsterAttributes.id, monsterDropID.intValue];
```
With this (and the similar changes below), are there any constants at the top of this file that aren't used at all anymore?

vokal-isaac commented on MH4U Dex/Images.xcassets/Region Icons/ancestral_steppe-icon.imageset/Contents.json:
```diff
@@ -0,0 +1,21 @@
+    {
+      "idiom" : "universal",
+      "scale" : "1x",
+      "filename" : "Image.png"
```
Did this image come from somewhere?  Are there any potential licensing issues with including it?

jmgoldb replied:
Thanks for this catch!  It came from a wiki with CC-BY-SA 3.0.  It's fine to include it, but I'll have to include an attributions page somewhere.

I think I should go through every image I have, and every Image I plan to add, and check the license on all of them.

vokal-isaac replied:
CC-BY-SA 3.0 includes this (SA) provision:

> ShareAlike — If you remix, transform, or build upon the material, you must distribute your contributions under the same license as the original.

which I would take to mean that you'd have to distribute this whole app as CC-BY-SA 3.0.

jmgoldb replied:
Oh!   Hrm...  

That could be a problem, I'll have to rethink that, or just leave the maps out in a worst-case scenario.

vokal-isaac replied:
Given that, it might be better to take the maps out for now.

jmgoldb replied:
A big issue will be the monster icons, though.  I don't think I can get away with leaving those out.  I'll have to figure something out.

vokal-isaac commented on MH4U Dex/ItemAreaSourcesTableViewController.m:
```diff
@@ -79,10 +79,12 @@ - (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(N
     UIStoryboard *storyBoard = [UIStoryboard storyboardWithName:@"Main" bundle:nil];
-    RegionEncyclopediaCollectionViewController *regionVC = [storyBoard instantiateViewControllerWithIdentifier:@"RegionEncyclopediaCollectionViewController"];
-    regionVC.managedObjectContext = self.managedObjectContext;
+    AreaDrop *drop = self.sources[indexPath.row];
```
With table view controllers like this, where you're accessing a property array, I often find it useful to declare helper method(s) to get the model object for a given index path—e.g., `- (AreaDrop *)dropAtIndexPath:(NSIndexPath *)indexPath`—to centralize that logic and to give you a correctly-typed result from the method call (rather than by assigning the item into a typed variable).

vokal-isaac commented on MH4U Dex/RegionContainerViewController.m:
```diff
@@ -0,0 +1,46 @@
+
+#import "CoreDataController.h"
+
+//NSString *const RegionNameKey = @"name";
```
Does this still need to be here?

vokal-isaac commented on MH4U Dex/RegionContainerViewController.m:
```diff
@@ -0,0 +1,46 @@
+- (void)viewDidLoad
+{
+    [super viewDidLoad];
+    //[self.regionMapSubview setHidden:YES];
```
Is this still needed?

jmgoldb replied:
I actually will later uncomment this, so this one I'll leave as-is.  The earlier one can go, though.

vokal-isaac replied:
Then add at least a comment-note to that effect and perhaps a `TODO:`, depending on why it's commented out and when/why it'll be re-added.

jmgoldb replied:
Will do.

vokal-isaac commented on MH4U Dex/RegionEncyclopediaCollectionViewController.m:
```diff
@@ -56,10 +59,27 @@ - (UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cell
+        regionVC.regionKeyName = [sender regionKeyName];
+        
+        CoreDataController *coreDataController = [CoreDataController sharedCDController];
+        NSPredicate *predicate = [NSPredicate predicateWithFormat:@"%K == %@", @"name", [sender regionName]];
```
Does mogenerator provide you a constant you can use in place of the `@"name"` string literal?

jmgoldb replied:
That would be RegionAttributes.name, yes?

vokal-isaac replied:
That seems likely from context, but you should be able to verify that by checking its value and verifying that it works.

vokal-isaac commented on MH4U Dex/RegionEncylcopediaCollectionViewCell.m:
```diff
@@ -8,6 +8,26 @@
+        self.regionIconImageView.image = image;
+    } else {
+        self.regionIconImageView.image = nil;
+    }
```
The `else` clause here is basically "If `image` is `nil`, set `self.regionIconImageView.image` to `nil`", which could just as easily be setting `self.regionIconImageView.image` to `image`—so you don't really need the `if`-`else`, unless there's something else coming here?

jmgoldb replied:
That's true, but for me at least, writing it this way makes it seem clear to me that leaving the image blank is unambiguously what should happen if the image is not found.

vokal-isaac commented on MH4U Dex/RegionItemTableViewController.m:
```diff
@@ -0,0 +1,87 @@
+    [areaFetchRequest setSortDescriptors:@[areaSortDescriptor]];
+    self.areaArray = [coreDataController.managedObjectContext executeFetchRequest:areaFetchRequest error:&fetchError];
+    
+   
```
Extra blank line(s).

vokal-isaac commented on MH4U Dex/RegionItemTableViewController.m:
```diff
@@ -0,0 +1,87 @@
+- (void)viewDidLoad
+{
+    CoreDataController *coreDataController = [CoreDataController sharedCDController];
+    self.areaDrops = [NSMutableDictionary dictionary];
```
It doesn't look like `self.areaDrops` gets used anywhere.

vokal-isaac commented on MH4U Dex/RegionItemTableViewController.m:
```diff
@@ -0,0 +1,87 @@
+    NSFetchRequest *areaFetchRequest = [[NSFetchRequest alloc] initWithEntityName:[Area entityName]];
+    NSPredicate *areaPredicate = [NSPredicate predicateWithFormat:@"%K == %@", @"region", self.region];
+    //TODO: Support all ranks.
+    NSPredicate *rankPredicate = [NSPredicate predicateWithFormat:@"%K == %@", @"rank", @"Low"];
```
Could you use mogenerator-provided constants instead of the string literal here and two lines up and two more lines up?

vokal-isaac commented on MH4U Dex/RegionItemTableViewController.m:
```diff
@@ -0,0 +1,87 @@
+{
+    Area *area = self.areaArray[indexPath.section];
+    NSSortDescriptor *dropSortDescriptor = [NSSortDescriptor sortDescriptorWithKey:@"idDecimalString" ascending:YES];
+    NSArray *dropsArray = [area.drop sortedArrayUsingDescriptors:@[dropSortDescriptor]];
```
This may lead to performance issues.

jmgoldb replied:
You're right, I should find some way to reorganize things so this is all set up in advance in viewDidLoad, or something.

vokal-isaac commented on MH4U Dex/RegionItemTableViewController.m:
```diff
@@ -0,0 +1,87 @@
+    CoreDataController *coreDataController = [CoreDataController sharedCDController];
+    itemVC.managedObjectContext = coreDataController.managedObjectContext;
+    Area *area = self.areaArray[indexPath.section];
+    NSSortDescriptor *dropSortDescriptor = [NSSortDescriptor sortDescriptorWithKey:@"idDecimalString" ascending:YES];
```
Is there a constant you could use in place of this string literal?

jmgoldb committed:
Added some method documentation comments, and removed images.  Also removed some NSString constants that were no longer being used.

jmgoldb committed:
Removed sorting from cellForRowAtIndexPath, will remove it from didSelectRowAtIndexPath next

jmgoldb committed:
Removed sorting from didSelectRowAtIndexPath

jmgoldb committed:
Support for Low, High, and G-Rank regions

jmgoldb committed:
Implemented Low-High-G rank monster drop tables

jmgoldb committed:
Quest Data is now loaded in.
Still need to create mock data for quest rewards, but first, I need to optimize loading speed for areas/regions, which takes ~30 seconds (far too long)

jmgoldb committed:
Quest drop data is loaded in

jmgoldb committed:
Addressed all outstanding code review issues, and made progress towards implementing display of quests to user.

jmgoldb commented:
I think I've addressed all the remaining pull request issues.

vokal-isaac commented on MH4U Dex/RegionItemTableViewController.m:
```diff
@@ -0,0 +1,107 @@
+    self.areaArray = [coreDataController.managedObjectContext executeFetchRequest:areaFetchRequest error:&fetchError];
+    self.areaDropsDict = [NSMutableDictionary dictionary];
+    for (Area *area in self.areaArray) {
+        NSSortDescriptor *dropSortDescriptor = [NSSortDescriptor sortDescriptorWithKey:AreaDropAttributes.idDecimalString ascending:YES];
```
Since this doesn't depend on the specific value of `area`, it would be more efficient to move it outside the loop.

vokal-isaac commented on MH4U Dex/RegionItemTableViewController.m:
```diff
@@ -0,0 +1,107 @@
+    self.areaDropsDict = [NSMutableDictionary dictionary];
+    for (Area *area in self.areaArray) {
+        NSSortDescriptor *dropSortDescriptor = [NSSortDescriptor sortDescriptorWithKey:AreaDropAttributes.idDecimalString ascending:YES];
+        NSArray *dropsArray = [area.drop sortedArrayUsingDescriptors:@[dropSortDescriptor]];
```
Actually, you might create the array `@[dropSortDescriptor]` outside the loop, too.

vokal-isaac commented on MH4U Dex/MonsterDetailsViewController.m:
```diff
@@ -29,39 +34,61 @@ @implementation MonsterDetailsViewController
+            self.overviewSubView.hidden = YES;
+            self.lowRankSubView.hidden = YES;
+            self.highRankSubView.hidden = YES;
+            self.GRankSubView.hidden = NO;
```
Missing `break` for this case.

vokal-isaac commented on MH4U Dex/MonsterDetailsViewController.m:
```diff
@@ -29,39 +34,61 @@ @implementation MonsterDetailsViewController
+            self.overviewSubView.hidden = YES;
+            self.lowRankSubView.hidden = YES;
+            self.highRankSubView.hidden = NO;
+            self.GRankSubView.hidden = YES;
```
Missing `break` for this case.

vokal-isaac commented on MH4U Dex/MonsterDetailsViewController.m:
```diff
@@ -29,39 +34,61 @@ @implementation MonsterDetailsViewController
+            self.GRankSubView.hidden = NO;
         default:
             break;
     }
```
It might be more concise to set all four `.hidden` properties to `YES` above the `switch`-`case` statement, so that you only have to set one `.hidden` property to `NO` in each case.

vokal-isaac commented on MH4U Dex/MonsterDetailsViewController.m:
```diff
@@ -29,39 +34,61 @@ @implementation MonsterDetailsViewController
+            self.lowRankSubView.hidden = NO;
+            self.highRankSubView.hidden = YES;
+            self.GRankSubView.hidden = YES;
             self.navigationItem.title = [NSString stringWithFormat:@"%@ LR Drops", self.monster];
```
Check into the difference between setting `self.navigationItem.title` and `self.title` to make sure that you're setting the correct one.

jmgoldb replied:
Changed these, but setting self.navigationItem.title has the same result as setting self.title.  I'm not sure why that is.

vokal-isaac commented on MH4U Dex/MonsterItemTableViewController.m:
```diff
@@ -1,45 +1,56 @@
+        rankPredicate = [NSPredicate predicateWithFormat:@"%K == %@", MonsterDropAttributes.rank, self.rank];
+    } else {
+        rankPredicate = [NSPredicate predicateWithFormat:@"%K == %@", MonsterDropAttributes.rank, @"Low"];
+    }
```
You could rewrite these 5 lines as:

``` objc
rankPredicate = [NSPredicate predicateWithFormat:@"%K == %@", MonsterDropAttributes.rank, self.rank ?: @"Low"];
```

The `?:` operator returns the first input if it's not `nil`/false-y, otherwise it returns the second input.

Is there a constant for `@"Low"`?  If not, would it make sense to define one?

vokal-isaac commented on MH4U Dex/MonsterDetailsViewController.m:
```diff
@@ -29,39 +34,61 @@ @implementation MonsterDetailsViewController
+    self.overviewSubView.hidden = NO;
+    self.lowRankSubView.hidden = YES;
+    self.highRankSubView.hidden = YES;
+    self.GRankSubView.hidden = YES;
```
Could you call `[self segmentedControlChanged:]` here rather than duplicating what it does for one case?

jmgoldb replied:
How would I do this without a segmented control to pass as a parameter?

vokal-isaac replied:
Have an `IBOutlet` pointed at the segmented control and pad that property as the sender.

vokal-isaac commented on MH4U Dex/MonsterDetailsViewController.m:
```diff
@@ -29,39 +34,61 @@ @implementation MonsterDetailsViewController
+    } else if ([segue.identifier isEqualToString:@"embedGRank"]) {
+        MonsterItemTableViewController *GRankVC = (MonsterItemTableViewController *)segue.destinationViewController;
+        GRankVC.managedObjectContext = self.managedObjectContext;
+        GRankVC.monster = self.monster;
```
These three lines are basically identical, except for the variable name, in each block of this `if`-`else`-`if` chain—you could probably move some form of it out to before the `if`s or otherwise restructure the `if`s to avoid the repetition.

vokal-isaac commented on MH4U Dex/MonsterItemTableViewController.m:
```diff
@@ -1,45 +1,56 @@
+    }
+    NSCompoundPredicate *compoundPredicate = [NSCompoundPredicate andPredicateWithSubpredicates:@[dropPredicate, rankPredicate]];
+    [fetchRequest setPredicate:compoundPredicate];
     [fetchRequest setSortDescriptors:@[sortDescriptor]];
```
`predicate` and `sortDescriptors` are properties, so you should use dot notation.

vokal-isaac commented on MH4U Dex/CoreDataController.m:
```diff
@@ -26,53 +30,87 @@ @interface CoreDataController ()
 
-NSString *const JSON = @"json";
+static NSString *const JSON = @"json";
+static NSString *const FalseString = @"FALSE";
```
Is this used anywhere?

vokal-isaac commented on MH4U Dex/CoreDataController.m:
```diff
@@ -177,6 +210,50 @@ - (void)loadMonsterDropData
+    for (NSDictionary *questDict in questList) {
+        Quest *quest;
+        NSNumber *questID = questDict[QuestIDKey];
+        NSPredicate *questPredicate = [NSPredicate predicateWithFormat:@"%K == %d", QuestAttributes.id, questID.intValue];
```
It would be better to use `questID` and `%@`—using `.intValue` and `%d` limits the range of values that will work more than using the `NSNumber` object directly.

vokal-isaac commented on MH4U Dex/CoreDataController.m:
```diff
@@ -363,6 +439,87 @@ - (NSUInteger)countForEntityWithEntityName:(NSString *)entityName withPredicate:
+    Quest *quest = [Quest insertInManagedObjectContext:self.managedObjectContext];
+    quest.id = questDict[QuestIDKey];
+    quest.name = questDict[QuestNameKey];
+    if ([questDict[QuestCaravanHallKey] isEqualToString:QuestCaravanKey]) {
```
Why are these coming through as `NSString` objects in the JSON rather than `NSNumber` objects with boolean values?

jmgoldb replied:
It's related to how I was/am generating the data, it's easier and faster for me to generate it that way.

vokal-isaac commented on MH4U Dex/CoreDataController.m:
```diff
@@ -363,6 +439,87 @@ - (NSUInteger)countForEntityWithEntityName:(NSString *)entityName withPredicate:
+            monsterPredicate = [NSPredicate predicateWithFormat:@"%K == %@", MonsterAttributes.name, questDict[QuestThirdMonsterKey]];
+            quest.thirdMonster = (Monster *)[self getUniqueEntityWithEntityName:[Monster entityName] withPredicate:monsterPredicate];
+            if (quest.thirdMonster) {
+                monsterPredicate = [NSPredicate predicateWithFormat:@"%K == %@", MonsterAttributes.name, questDict[QuestFourthMonsterKey]];
```
It may be worth creating a method on, say, the `Monster` class along the lines of `predicateForMonsterWithName:(NSString *)name` to make these predicate-creation lines shorter and less repetitive.

vokal-isaac replied:
... or maybe better, a method like `getMonsterWithName:(NSString *)name` or something.

vokal-isaac commented on MH4U Dex/Model/Private/_Monster.h:
```diff
@@ -17,10 +17,18 @@ extern const struct MonsterAttributes {
+@class Quest;
+@class Quest;
+@class Quest;
+@class Quest;
```
<s>Is there a reason this is repeated four times?</s>  ... oh, nevermind, this is mogenerator-generated.

vokal-isaac commented on MH4U Dex/CoreDataController.m:
```diff
@@ -177,6 +210,50 @@ - (void)loadMonsterDropData
+        QuestDrop *drop;
+        NSNumber *dropID = dropDict[QuestDropIDKey];
+        NSPredicate *dropPredicate = [NSPredicate predicateWithFormat:@"%K == %d", QuestDropAttributes.id, dropID.intValue];
+        drop = (QuestDrop *)[self getUniqueEntityWithEntityName:[QuestDrop entityName] withPredicate:dropPredicate];
```
You could combine the `QuestDrop *drop;` line with this line.

vokal-isaac commented on MH4U Dex/CoreDataController.m:
```diff
@@ -177,6 +210,50 @@ - (void)loadMonsterDropData
+        if (!drop) {
+            drop = [self questDropFromDictionary:dropDict];
+            if (!drop) {
+                //If we reach this point, then the quest that this drop belongs to did not exist, and we should return without saving the context;
```
If I'm reading it correctly, it's not just that it did not exist, it's that you tried to create it and failed to do so...

jmgoldb replied:
I think you might be reading this incorrectly, let's talk about it later today.

vokal-isaac commented on MH4U Dex/CoreDataController.m:
```diff
@@ -363,6 +439,87 @@ - (NSUInteger)countForEntityWithEntityName:(NSString *)entityName withPredicate:
+        //If the quest was not found, then it makes no sense to store this reward.  Just return nil.
+        NSLog(@"The quest for a quest reward was not found!");
+        return nil;
+    }
```
You will still have created the `QuestDrop` object `drop`, though—you might want to try to retrieve the quest first, check that it exists, then create `drop`.

vokal-isaac commented on MH4U Dex/QuestEncyclopediaContainerViewController.m:
```diff
@@ -0,0 +1,38 @@
+- (IBAction)segmentedControlChanged:(id)sender
+{
+    
+}
```
Is this intentionally empty?  Should there be a `// TODO:` comment in here?

vokal-isaac commented on MH4U Dex/QuestEncyclopediaContainerViewController.m:
```diff
@@ -0,0 +1,38 @@
+        QuestListTableViewController *questListVC = (QuestListTableViewController *)segue.destinationViewController;
+        questListVC.isCaravan = YES;
+    } else if ([segue.identifier isEqualToString:@"showGatheringHallList"]) {
+        QuestListTableViewController *questListVC = (QuestListTableViewController *)segue.destinationViewController;
```
This line is in both blocks above—would it make sense to pull it above the `if` statement?

vokal-isaac commented on MH4U Dex/QuestListTableViewController.m:
```diff
@@ -0,0 +1,89 @@
+    [super viewDidLoad];
+    
+    
+    
```
Extra blank lines.

vokal-isaac commented on MH4U Dex/QuestListTableViewController.m:
```diff
@@ -0,0 +1,89 @@
+    
+    
+    
+    CoreDataController *coreDataController = [CoreDataController sharedCDController];
```
It looks like you're only using this for the `.managedObjectContext` property—if so, it might make sense to have an `NSManagedObjectContext` variable and store the value of that property in it straight away, rather than storing the `CoreDataController`.

vokal-isaac commented on MH4U Dex/QuestListTableViewController.m:
```diff
@@ -0,0 +1,89 @@
+        starPredicate = [NSPredicate predicateWithFormat:@"%K == %@", QuestAttributes.stars, @(starNumber)];
+        compoundPredicate = [NSCompoundPredicate andPredicateWithSubpredicates:@[caravanPredicate, starPredicate]];
+        [questFetchRequest setPredicate:compoundPredicate];
+        [self.questMasterArray addObject:[coreDataController.managedObjectContext executeFetchRequest:questFetchRequest error:&fetchError]];;
```
`executeFetchRequest:error:` might return `nil`, which would cause a crash for attempting to insert `nil` into an array.  You might also want to do something with `fetchError` when it returns `nil`.

vokal-isaac replied:
Extra `;`.

vokal-isaac commented on MH4U Dex/RegionEncyclopediaCollectionViewController.m:
```diff
@@ -8,12 +8,17 @@
 
+#import "RegionEncylcopediaCollectionViewCell.h"
+
+
```
Extra blank line.

vokal-isaac commented on MH4U Dex/CoreDataController.m:
```diff
@@ -93,33 +131,33 @@ - (void)loadMonsterData
+                   @"ice": @"ice",
+                   @"dragon": @"dragon",
+                   @"ko": @"ko",
+                   @"extract": @"extract",
```
Do constants exist for any of these strings?  Perhaps mogenerator-generated attribute names?

vokal-isaac commented on MH4U Dex/RegionEncylcopediaCollectionViewCell.m:
```diff
@@ -8,6 +8,26 @@
 
+- (void)displayContents
+{
+    NSString *imageName = [NSString stringWithFormat:@"%@-icon", self.regionKeyName];
```
``` objc
NSString *imageName = [self.regionKeyName stringByAppendingString:@"-icon"];
```

Also, you've got basically this same construct in two far-apart places, so it might make sense to consolidate this logic in one more-centralized place.

vokal-isaac commented on MH4U Dex/MonsterEncyclopediaViewController.m:
```diff
@@ -37,8 +39,8 @@ - (void)viewDidLoad
-    NSSortDescriptor *sortDescriptor = [NSSortDescriptor sortDescriptorWithKey:@"sort_name" ascending:YES];
-    NSPredicate *predicate = [NSPredicate predicateWithFormat:@"%K == %@", @"monster_class", @"Minion"];
+    NSSortDescriptor *sortDescriptor = [NSSortDescriptor sortDescriptorWithKey:MonsterAttributes.sort_name ascending:YES];
+    NSPredicate *predicate = [NSPredicate predicateWithFormat:@"%K == %@", MonsterAttributes.monster_class, @"Minion"];
```
Are there constants for `@"Minion"` here and `@"Boss"` below?

vokal-isaac commented on MH4U Dex/RegionItemTableViewController.m:
```diff
@@ -0,0 +1,107 @@
+        rankPredicate = [NSPredicate predicateWithFormat:@"%K == %@", AreaAttributes.rank, self.rank];
+    } else {
+        rankPredicate = [NSPredicate predicateWithFormat:@"%K == %@", AreaAttributes.rank, @"Low"];
+    }
```
I think I've seen these 5 lines somewhere else...

vokal-isaac commented on MH4U Dex/MonsterItemTableViewController.h:
```diff
@@ -8,9 +8,10 @@
+@interface MonsterItemTableViewController : UITableViewController
 
 @property (nonatomic, strong) NSManagedObjectContext *managedObjectContext;
 @property (nonatomic, strong) NSString *monster;
```
I wouldn't expect a property named `monster` to be a string—should this be a `Monster` object or should it be called `monsterName` or something else?

vokal-isaac commented on MH4U Dex/MonsterOverviewViewController.h:
```diff
@@ -13,6 +13,4 @@
 @property (nonatomic, strong) NSManagedObjectContext *managedObjectContext;
 @property (nonatomic, strong) NSString *monster;
```
Ditto here.

vokal-isaac commented on MH4U Dex/DropTableViewCell.h:
```diff
@@ -15,6 +15,9 @@
+/**
+ Update the cell's labels.
+ */
 - (void)displayContents;
```
Should this method be passed a model object?

vokal-isaac commented on MH4U Dex/ItemMonsterSourceTableViewCell.h:
```diff
@@ -13,6 +13,9 @@
+/**
+ Update the cell's labels.
+ */
 - (void)displayContents;
```
Should this method be passed a model object?

vokal-isaac commented on MH4U Dex/ItemTableViewCell.h:
```diff
@@ -12,6 +12,10 @@
+/**
+ Update the cell's labels.
+ */
 - (void)displayContents;
```
Should this method be passed a model object?

vokal-isaac commented on MH4U Dex/MonsterEncyclopediaCell.h:
```diff
@@ -12,6 +12,9 @@
+/**
+ Update the cell's labels.
+ */
 - (void)displayContents;
```
Should this method be passed a model object?

vokal-isaac commented on MH4U Dex/QuestListTableViewCell.m:
```diff
@@ -0,0 +1,13 @@
+
+@implementation QuestListTableViewCell
+
+@end
```
Empty class?  Should there be a `// TODO:` comment?

vokal-isaac commented on MH4U Dex/RegionEncylcopediaCollectionViewCell.h:
```diff
@@ -10,4 +10,12 @@
+/**
+ Update the cell's labels.
+ */
+- (void)displayContents;
```
Should this method be passed a model object?

jmgoldb replied:
I could pass it a model object instead of storing the two strings as properties, if that would be preferable

seanwolter replied:
Our preferred pattern is to make a method like `layoutWithThing:` or `configureWithThing:` and pass, but not retain, the model object.

jmgoldb committed:
A lot of refactoring done to address code review comments.

jmgoldb committed:
changed how monsterView titlebar changes

jmgoldb commented:
I've addressed all of the issues, aside from declaring constants for strings that get used more than once, which I'd like to do by making an external constants file as the next pull request.  I also haven't consolidated the image handling into one place, I'm not sure yet where I want to put that.  Everything else should be accounted for.

vokal-isaac commented on MH4U Dex/CoreDataController.m:
```diff
@@ -363,6 +438,87 @@ - (NSUInteger)countForEntityWithEntityName:(NSString *)entityName withPredicate:
+- (QuestDrop *)questDropFromDictionary:(NSDictionary *)dropDict
+{
+    NSPredicate *questPredicate = [NSPredicate predicateWithFormat:@"%K == %@", QuestAttributes.id, dropDict[QuestDropQuestIDKey]];
+    if (![self getUniqueEntityWithEntityName:[Quest entityName] withPredicate:questPredicate]) {
```
To avoid having to run this `getUniqueEntityWithEntityName:withPredicate:` call twice, you could store it into a local variable, e.g.:

``` objc
Quest *quest = (Quest *)[self getUniqueEntityWithEntityName:[Quest entityName] withPredicate:questPredicate];
if (!quest) {
...
drop.quest = quest.
```

seanwolter replied:
I know I'm late to the game on this project. Please forgive these points:
1) Generally "get" as a prefix is verboten. I'd prefer "fetch" or no prefix for a method like this.
2) Why not use Vokoder? I respect the experienced gained building your own core data stack, but I'd also like you to be up to speed on our standard tools.

vokal-isaac commented on MH4U Dex/MonsterEncyclopediaViewController.m:
```diff
@@ -62,9 +64,26 @@ - (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender
+            return self.largeMonsters[indexPath.row];
+            break;
+    }
+    return nil;
```
You don't need the `break;`s after `return`s, but you can leave them in if you want.  Similarly, it's a little more typical to include a `default:` section and put the `return nil;` in there.

vokal-isaac commented on MH4U Dex/MonsterDetailsViewController.m:
```diff
@@ -29,43 +35,57 @@ @implementation MonsterDetailsViewController
+    } else {
+        MonsterItemTableViewController *itemVC = (MonsterItemTableViewController *)segue.destinationViewController;
+        itemVC.managedObjectContext = self.managedObjectContext;
+        itemVC.monsterName = self.monster;
```
Is this (and 4 lines farther up) another instance of a property named `monster` that's a name string rather than a `Monster` object?

vokal-isaac commented on MH4U Dex/QuestListTableViewController.m:
```diff
@@ -0,0 +1,92 @@
+    NSCompoundPredicate *compoundPredicate;
+    for (NSUInteger starNumber = 1; starNumber <= 10; starNumber++) {
+        starPredicate = [NSPredicate predicateWithFormat:@"%K == %@", QuestAttributes.stars, @(starNumber)];
+        compoundPredicate = [NSCompoundPredicate andPredicateWithSubpredicates:@[caravanPredicate, starPredicate]];
```
Can this line (and `NSCompoundPredicate *compoundPredicate;` above) be removed?

vokal-isaac commented on MH4U Dex/RegionEncyclopediaCollectionViewController.m:
```diff
@@ -53,13 +56,34 @@ - (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSe
+            Region *region = [self regionAtIndexPath:[self.collectionView indexPathForCell:cell]];
+            regionVC.regionName = region.name;
+            regionVC.regionKeyName = region.keyName;
+            regionVC.region = region;
```
If you're setting a `region` property, are the `regionName` and `regionKeyName` properties on `regionVC` necessary?

jmgoldb committed:
Addressed code review feedback

vokal-isaac commented:
Good to pull.

Keep at least a mental TODO to look at switching to Vokoder, as Sean suggested.

seanwolter commented:
Joe answered my question about Vokoder vs. build a core data castle from scratch.
Were the `get...` methods changed? Or does no one else mind that?

vokal-isaac commented:
I believe d8a5f0b renamed the `get...` method.
